### DISCUSSION TITLE
What exactly is a "frozen" processor / what is the purpose of freezing a processor?

### DISCUSSION THREAD

**Discussion by ben519** [unifiedjs/unified#222] [Q&A] [answered]
New to unifiedjs. Reading the docs, I see

> When processors are exposed from a module (for example, unified itself) they should not be configured directly, as that would change their behavior for all module users. Those processors are [frozen](https://www.npmjs.com/package/unified#processorfreeze) and they should be called to create a new processor before they are used.

Sorry, but this isn't really clicking for me. Perhaps someone could give an example of a frozen processor versus a non frozen processor? Why might I freeze (or not freeze) a processor? Thanks!

**Comment by wooorm** [comment]
Did you click that link? https://github.com/unifiedjs/unified#processorfreeze. That's some more info and a pretty thorough example?

**Reply by ben519** [reply]
Hey thanks! I had not seen that link. Can you please tell me if my understanding is correct?

In this example, we create a processor called `rehype`

```
export const rehype = unified().use(rehypeParse).use(rehypeStringify).freeze()
```

We can *configure* it

```
const myNewProcessor = rehype().use(rehypeFormat)
```

but we cannot *extend* it

```
const myNewProcessor = rehype.use(rehypeFormat)
```

Why exactly does this matter? Is it that `rehype().use(rehypeFormat)` traverses the tree twice while `rehype.use(rehypeFormat)` would only traverse the tree once? Thanks!

**Reply by wooorm** [reply, marked as answer]
```js
const myNewProcessor = rehype().use(rehypeFormat)
```

This *clones* (`()`), then *configures* (`.use(...)`)

```js
const myNewProcessor = rehype.use(rehypeFormat)
```

This tries to configure something that is shared with everyone. If you would be able to configure it, everyone else would also be configured.

The reason is that how *you* use `rehype` in one place, should not affect how other people, other places, use it!

Traversing trees happens somewhere else. This is unrelated to that!
